File: typescript/packages/sdk/src/albums.ts
import type { AlbumCreate, AlbumUpdate, AssetSelection, ShareLinkCreate } from '@imogen/shared'
import {
  Album,
  AlbumAssetsRemoved,
  AlbumAssetsResult,
  AlbumList,
  AlbumWithAssets,
  ShareLink,
  ShareLinkOrNone,
} from '@imogen/shared'
import { selectionBody } from './assets.js'
import type { HttpClient } from './http.js'

export class Albums {
  constructor(private readonly http: HttpClient) {}

  async list(): Promise<Album[]> {
    const { items } = await this.http.request('GET', '/api/v1/albums', { decode: AlbumList })
    return items
  }

  get(albumId: string): Promise<AlbumWithAssets> {
    return this.http.request('GET', `/api/v1/albums/${albumId}`, { decode: AlbumWithAssets })
  }

  create(input: AlbumCreate): Promise<Album> {
    return this.http.request('POST', '/api/v1/albums', { body: input, decode: Album })
  }

  update(albumId: string, patch: AlbumUpdate): Promise<Album> {
    return this.http.request('PATCH', `/api/v1/albums/${albumId}`, {
      body: patch,
      decode: Album,
    })
  }

  remove(albumId: string): Promise<void> {
    return this.http.request<void>('DELETE', `/api/v1/albums/${albumId}`)
  }

  addAssets(albumId: string, selection: string[] | AssetSelection): Promise<AlbumAssetsResult> {
    return this.http.request('POST', `/api/v1/albums/${albumId}/assets`, {
      body: selectionBody(selection),
      decode: AlbumAssetsResult,
    })
  }

  removeAssets(
    albumId: string,
    selection: string[] | AssetSelection,
  ): Promise<{ removed: number }> {
    return this.http.request('DELETE', `/api/v1/albums/${albumId}/assets`, {
      body: selectionBody(selection),
      decode: AlbumAssetsRemoved,
    })
  }

  /** The live public link for this album, or null. */
  shareLink(albumId: string): Promise<ShareLink | null> {
    return this.http.request('GET', `/api/v1/albums/${albumId}/share`, {
      decode: ShareLinkOrNone,
    })
  }

  share(albumId: string, input: ShareLinkCreate = { allowDownload: true }): Promise<ShareLink> {
    return this.http.request('POST', `/api/v1/albums/${albumId}/share`, {
      body: input,
      decode: ShareLink,
    })
  }

  unshare(albumId: string): Promise<void> {
    return this.http.request<void>('DELETE', `/api/v1/albums/${albumId}/share`)
  }
}

File: rust/src/albums.rs
use std::sync::Arc;

use reqwest::Method;
use serde_json::json;

use crate::error::Result;
use crate::http::{HttpClient, RequestOptions};
use crate::models::*;

pub struct Albums {
    http: Arc<HttpClient>,
}

impl Albums {
    pub(crate) fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn list(&self) -> Result<Vec<Album>> {
        let page: Items<Album> = self
            .http
            .request(Method::GET, "/api/v1/albums", RequestOptions::default())
            .await?;
        Ok(page.items)
    }

    pub async fn get(&self, album_id: &str) -> Result<AlbumWithAssets> {
        self.http
            .request(
                Method::GET,
                &format!("/api/v1/albums/{album_id}"),
                RequestOptions::default(),
            )
            .await
    }

    pub async fn create(&self, input: &AlbumCreate) -> Result<Album> {
        self.http
            .request(Method::POST, "/api/v1/albums", RequestOptions::json(input)?)
            .await
    }

    pub async fn update(&self, album_id: &str, patch: &AlbumUpdate) -> Result<Album> {
        self.http
            .request(
                Method::PATCH,
                &format!("/api/v1/albums/{album_id}"),
                RequestOptions::json(patch)?,
            )
            .await
    }

    pub async fn remove(&self, album_id: &str) -> Result<()> {
        self.http
            .request::<Option<serde_json::Value>>(
                Method::DELETE,
                &format!("/api/v1/albums/{album_id}"),
                RequestOptions::default(),
            )
            .await?;
        Ok(())
    }

    pub async fn add_assets(
        &self,
        album_id: &str,
        asset_ids: &[String],
    ) -> Result<AlbumAssetsResult> {
        self.http
            .request(
                Method::POST,
                &format!("/api/v1/albums/{album_id}/assets"),
                RequestOptions::json(&json!({ "assetIds": asset_ids }))?,
            )
            .await
    }

    pub async fn remove_assets(
        &self,
        album_id: &str,
        asset_ids: &[String],
    ) -> Result<RemovedCount> {
        self.http
            .request(
                Method::DELETE,
                &format!("/api/v1/albums/{album_id}/assets"),
                RequestOptions::json(&json!({ "assetIds": asset_ids }))?,
            )
            .await
    }

    /// The live public link for this album, or `None`.
    pub async fn share_link(&self, album_id: &str) -> Result<Option<ShareLink>> {
        self.http
            .request(
                Method::GET,
                &format!("/api/v1/albums/{album_id}/share"),
                RequestOptions::default(),
            )
            .await
    }

    pub async fn share(&self, album_id: &str, input: &ShareLinkCreate) -> Result<ShareLink> {
        self.http
            .request(
                Method::POST,
                &format!("/api/v1/albums/{album_id}/share"),
                RequestOptions::json(input)?,
            )
            .await
    }

    pub async fn unshare(&self, album_id: &str) -> Result<()> {
        self.http
            .request::<Option<serde_json::Value>>(
                Method::DELETE,
                &format!("/api/v1/albums/{album_id}/share"),
                RequestOptions::default(),
            )
            .await?;
        Ok(())
    }
}
